mqtt/mqtt-broker/src/server.rs

```diff
@@ -10,46 +11,54 @@ use tracing_futures::Instrument;
 use crate::auth::{Authenticator, Authorizer};
 use crate::broker::{Broker, BrokerHandle, BrokerState};
 use crate::transport::TransportBuilder;
-use crate::{connection, Error, InitializeBrokerError, Message, SystemEvent};
+use crate::{connection, AuthenticationError, Error, InitializeBrokerError, Message, SystemEvent};
 
-pub struct Server<N, Z>
+pub struct Server<Z>
 where
-    N: Authenticator,
     Z: Authorizer,
 {
-    broker: Broker<N, Z>,
+    broker: Broker<Z>,
 }
 
-impl<N, Z> Server<N, Z>
+impl<Z> Server<Z>
 where
-    N: Authenticator + Send + Sync + 'static,
     Z: Authorizer + Send + Sync + 'static,
 {
-    pub fn from_broker(broker: Broker<N, Z>) -> Self {
+    pub fn from_broker(broker: Broker<Z>) -> Self {
         Self { broker }
     }
 
-    pub async fn serve<A, F, I>(
+    pub async fn serve<A, F, I, N>(
         self,
         transports: I,
         shutdown_signal: F,
+        authenticator: N,
     ) -> Result<BrokerState, Error>
     where
         A: ToSocketAddrs,
         F: Future<Output = ()> + Unpin,
         I: IntoIterator<Item = TransportBuilder<A>>,
+        N: Authenticator + Send + Sync + 'static,
+        N::Error: Into<AuthenticationError>,
     {
         let Server { broker } = self;
         let mut handle = broker.handle();
         let broker_task = tokio::spawn(broker.run());
 
+        let authenticator = Arc::new(authenticator);
```

Comment:
just a question: It might be a good idea to use Arc instead of cloning. Just to confirm - when we actually call `authenticator.authenticate()`, it is not a blocking call, right?